File: src/api/bitrise.ts
import { fetch as tauriFetch } from '@tauri-apps/plugin-http';
import { BaseDirectory, writeFile, mkdir, exists } from '@tauri-apps/plugin-fs';
import { homeDir } from '@tauri-apps/api/path';
import { App, Build, Artifact } from '../types';

const BITRISE_API_BASE = 'https://api.bitrise.io/v0.1';

export class BitriseClient {
  private readonly token: string;

  constructor(token: string) {
    this.token = token;
  }

  async getApps(): Promise<App[]> {
    const url = `${BITRISE_API_BASE}/apps`;
    console.log('[BitriseClient] Fetching apps from:', url);
    console.log('[BitriseClient] Token (first 10 chars):', this.token.substring(0, 10) + '...');

    try {
      const response = await tauriFetch(url, {
        method: 'GET',
        headers: {
          'Authorization': `token ${this.token}`,
        },
      });

      console.log('[BitriseClient] Response status:', response.status);
      console.log('[BitriseClient] Response ok:', response.ok);
      console.log('[BitriseClient] Response headers:', response.headers);

      if (!response.ok) {
        const errorBody = await response.text();
        console.error('[BitriseClient] Error response body:', errorBody);
        throw new Error(`Failed to fetch apps: ${response.status} - ${errorBody}`);
      }

      const data = await response.json() as { data: App[] };
      console.log('[BitriseClient] Apps received:', data.data.length);
      console.log('[BitriseClient] First app:', data.data[0]);
      return data.data;
    } catch (error) {
      console.error('[BitriseClient] Exception in getApps:', error);
      console.error('[BitriseClient] Error type:', typeof error);
      console.error('[BitriseClient] Error details:', JSON.stringify(error, null, 2));
      throw error;
    }
  }

  async getBuilds(appSlug: string, branch?: string, next?: string): Promise<{ builds: Build[]; next?: string }> {
    let url = `${BITRISE_API_BASE}/apps/${appSlug}/builds?limit=50`;
    if (branch) {
      url += `&branch=${encodeURIComponent(branch)}`;
    }
    if (next) {
      url += `&next=${encodeURIComponent(next)}`;
    }
    console.log('[BitriseClient] Fetching builds from:', url);

    const response = await tauriFetch(url, {
      method: 'GET',
      headers: {
        'Authorization': `token ${this.token}`,
      },
    });

    if (!response.ok) {
      const body = await response.text();
      console.error('[BitriseClient] Builds error response:', body);
      throw new Error(`API error ${response.status}: ${body}`);
    }

    const data = await response.json() as { data: any[]; paging?: { next?: string } };
    console.log('[BitriseClient] Successfully parsed', data.data.length, 'builds');

    // Map triggered_workflow to workflow (Bitrise API returns triggered_workflow)
    const builds: Build[] = data.data.map((build: any) => ({
      ...build,
      workflow: build.workflow || build.triggered_workflow,
    }));

    return { builds, next: data.paging?.next };
  }

  async getArtifacts(appSlug: string, buildSlug: string): Promise<Artifact[]> {
    const url = `${BITRISE_API_BASE}/apps/${appSlug}/builds/${buildSlug}/artifacts`;
    console.log('[BitriseClient] Fetching artifacts from:', url);

    const response = await tauriFetch(url, {
      method: 'GET',
      headers: {
        'Authorization': `token ${this.token}`,
      },
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch artifacts: ${response.status}`);
    }

    const data = await response.json() as { data: Artifact[] };
    console.log('[BitriseClient] Artifacts received:', data.data.length);
    return data.data;
  }

  async downloadArtifact(
    appSlug: string,
    buildSlug: string,
    artifactSlug: string,
    fileName: string
  ): Promise<string> {
    // Get cache directory path
    const cacheDirName = '.bitrise-artifacts/cache';

    // Ensure cache directory exists
    try {
      const cacheExists = await exists(cacheDirName, { baseDir: BaseDirectory.Home });
      if (!cacheExists) {
        console.log('[BitriseClient] Creating cache directory:', cacheDirName);
        await mkdir(cacheDirName, { baseDir: BaseDirectory.Home, recursive: true });
      }
    } catch (error) {
      console.error('[BitriseClient] Error checking/creating cache dir:', error);
      throw new Error(`Failed to create cache directory: ${error}`);
    }

    const cacheFilePath = `${cacheDirName}/${fileName}`;
    console.log('[BitriseClient] Downloading artifact to cache:', cacheFilePath);

    // Step 1: Get artifact metadata with expiring download URL
    const metadataUrl = `${BITRISE_API_BASE}/apps/${appSlug}/builds/${buildSlug}/artifacts/${artifactSlug}`;
    const metadataResponse = await tauriFetch(metadataUrl, {
      method: 'GET',
      headers: {
        'Authorization': `token ${this.token}`,
      },
    });

    if (!metadataResponse.ok) {
      throw new Error(`Failed to get artifact info: ${metadataResponse.status}`);
    }

    const metadataJson = await metadataResponse.json() as {
      data: { expiring_download_url?: string }
    };
    const downloadUrl = metadataJson.data.expiring_download_url;

    if (!downloadUrl) {
      throw new Error('No download URL found in artifact metadata');
    }

    console.log('[BitriseClient] Download URL obtained, fetching artifact...');

    // Step 2: Download the actual file
    const artifactResponse = await tauriFetch(downloadUrl, {
      method: 'GET',
    });

    if (!artifactResponse.ok) {
      throw new Error(`Failed to download artifact: ${artifactResponse.status}`);
    }

    const bytes = new Uint8Array(await artifactResponse.arrayBuffer());
    console.log('[BitriseClient] Writing', bytes.length, 'bytes to', cacheFilePath);

    // Step 3: Write to cache directory
    await writeFile(cacheFilePath, bytes, { baseDir: BaseDirectory.Home });
    console.log('[BitriseClient] Download complete!');

    // Step 4: Construct and return the absolute path
    // The Rust commands (installApk, saveToDownloads) expect absolute paths
    const home = await homeDir();
    const absolutePath = `${home}/.bitrise-artifacts/cache/${fileName}`;
    console.log('[BitriseClient] Absolute path:', absolutePath);

    return absolutePath;
  }
}

File: src-tauri/src/bitrise.rs
use reqwest::Client;
use serde::{Deserialize, Serialize};

const BITRISE_API_BASE: &str = "https://api.bitrise.io/v0.1";

#[derive(Debug, Clone)]
pub struct BitriseClient {
    client: Client,
    token: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct App {
    pub slug: String,
    pub title: String,
    pub repo_slug: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Build {
    #[serde(default)]
    pub slug: Option<String>,
    #[serde(default)]
    pub build_number: Option<i64>,
    #[serde(default)]
    pub branch: Option<String>,
    #[serde(default)]
    pub status_text: Option<String>,
    #[serde(default)]
    pub triggered_at: Option<String>,
    #[serde(default)]
    pub finished_at: Option<String>,
    #[serde(default)]
    pub triggered_by: Option<String>,
    #[serde(default)]
    pub commit_message: Option<String>,
    #[serde(default)]
    pub commit_hash: Option<String>,
    #[serde(default)]
    pub pull_request_id: Option<i64>,
    #[serde(default)]
    pub pull_request_target_branch: Option<String>,
    #[serde(default)]
    pub pull_request_author: Option<String>,
    #[serde(default)]
    pub pipeline_id: Option<String>,
    #[serde(default)]
    pub pipeline_workflow_id: Option<String>,
    #[serde(default, alias = "triggered_workflow")]
    pub workflow: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Artifact {
    #[serde(default)]
    pub slug: Option<String>,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub artifact_type: Option<String>,
    #[serde(default)]
    pub file_size_bytes: Option<i64>,
    #[serde(default)]
    pub is_public_page_enabled: Option<bool>,
    #[serde(default)]
    pub public_install_page_url: Option<String>,
}

impl BitriseClient {
    pub fn new(token: String) -> Self {
        Self {
            client: Client::new(),
            token,
        }
    }

    pub async fn get_apps(&self) -> anyhow::Result<Vec<App>> {
        let url = format!("{}/apps", BITRISE_API_BASE);
        let response = self.client
            .get(&url)
            .header("Authorization", format!("token {}", self.token))
            .send()
            .await?;

        if !response.status().is_success() {
            anyhow::bail!("Failed to fetch apps: {}", response.status());
        }

        let data: serde_json::Value = response.json().await?;
        let apps: Vec<App> = serde_json::from_value(data["data"].clone())?;
        Ok(apps)
    }

    pub async fn get_builds(&self, app_slug: &str) -> anyhow::Result<Vec<Build>> {
        let url = format!("{}/apps/{}/builds?limit=50", BITRISE_API_BASE, app_slug);
        println!("Fetching builds from: {}", url);
        
        let response = self.client
            .get(&url)
            .header("Authorization", format!("token {}", self.token))
            .send()
            .await?;

        let status = response.status();
        println!("Builds response status: {}", status);
        
        if !status.is_success() {
            let body = response.text().await.unwrap_or_default();
            println!("Builds error response: {}", body);
            anyhow::bail!("API error {}: {}", status, body);
        }

        let data: serde_json::Value = response.json().await?;
        
        // Debug: print full first build to see actual structure
        if let Some(builds_array) = data["data"].as_array() {
            if let Some(first_build) = builds_array.first() {
                println!("\n=== FIRST BUILD DATA ===");
                println!("{}", serde_json::to_string_pretty(first_build).unwrap_or_default());
                println!("========================\n");
            }
        }
        
        let builds: Vec<Build> = serde_json::from_value(data["data"].clone())
            .map_err(|e| {
                println!("Failed to parse builds: {:?}", e);
                e
            })?;
        println!("Successfully parsed {} builds", builds.len());
        if let Some(first) = builds.first() {
            println!("First build workflow: {:?}", first.workflow);
        }
        Ok(builds)
    }

    pub async fn get_artifacts(&self, app_slug: &str, build_slug: &str) -> anyhow::Result<Vec<Artifact>> {
        let url = format!(
            "{}/apps/{}/builds/{}/artifacts",
            BITRISE_API_BASE, app_slug, build_slug
        );
        let response = self.client
            .get(&url)
            .header("Authorization", format!("token {}", self.token))
            .send()
            .await?;

        if !response.status().is_success() {
            anyhow::bail!("Failed to fetch artifacts: {}", response.status());
        }

        let data: serde_json::Value = response.json().await?;
        let artifacts: Vec<Artifact> = serde_json::from_value(data["data"].clone())?;
        Ok(artifacts)
    }

    pub async fn download_artifact(
        &self,
        app_slug: &str,
        build_slug: &str,
        artifact_slug: &str,
        file_name: &str,
    ) -> anyhow::Result<String> {
        // Get the cache directory
        let cache_dir = crate::fs::get_cache_dir()?;
        
        // Create full path
        let save_path = cache_dir.join(file_name);
        let save_path_str = save_path.to_string_lossy().to_string();
        
        println!("Downloading artifact to cache: {}", save_path_str);
        
        let url = format!(
            "{}/apps/{}/builds/{}/artifacts/{}",
            BITRISE_API_BASE, app_slug, build_slug, artifact_slug
        );
        let response = self.client
            .get(&url)
            .header("Authorization", format!("token {}", self.token))
            .send()
            .await?;

        if !response.status().is_success() {
            anyhow::bail!("Failed to get artifact info: {}", response.status());
        }

        let data: serde_json::Value = response.json().await?;
        let download_url = data["data"]["expiring_download_url"]
            .as_str()
            .ok_or_else(|| anyhow::anyhow!("No download URL found"))?;

        println!("Download URL obtained, fetching artifact...");
        let artifact_response = self.client.get(download_url).send().await?;
        let bytes = artifact_response.bytes().await?;
        
        println!("Writing {} bytes to {}", bytes.len(), save_path_str);
        tokio::fs::write(&save_path, bytes).await?;
        println!("Download complete!");
        
        Ok(save_path_str)
    }
}